packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/media/ImageStreamReader.java

```diff
@@ -0,0 +1,223 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package io.flutter.plugins.camera.media;
+
+import android.graphics.ImageFormat;
+import android.media.Image;
+import android.media.ImageReader;
+import android.os.Handler;
+import android.os.Looper;
+import android.view.Surface;
+import androidx.annotation.NonNull;
+import androidx.annotation.VisibleForTesting;
+import io.flutter.plugin.common.EventChannel;
+import io.flutter.plugins.camera.types.CameraCaptureProperties;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+// Wraps an ImageReader to allow for testing of the image handler.
+public class ImageStreamReader {
+
+  /**
+   * The image format we are going to send back to dart. Usually it's the same as streamImageFormat
+   * but in the case of NV21 we will actually request YUV frames but convert it to NV21 before
+   * sending to dart.
+   */
+  private final int dartImageFormat;
+
+  private final ImageReader imageReader;
+  private final ImageStreamReaderUtils imageStreamReaderUtils;
+
+  /**
+   * Creates a new instance of the {@link ImageStreamReader}.
+   *
+   * @param imageReader is the image reader that will receive frames
+   * @param imageStreamReaderUtils is an instance of {@link ImageStreamReaderUtils}
+   */
+  @VisibleForTesting
+  public ImageStreamReader(
+      ImageReader imageReader, int dartImageFormat, ImageStreamReaderUtils imageStreamReaderUtils) {
+    this.imageReader = imageReader;
+    this.dartImageFormat = dartImageFormat;
+    this.imageStreamReaderUtils = imageStreamReaderUtils;
+  }
+
+  /**
+   * Creates a new instance of the {@link ImageStreamReader}.
+   *
+   * @param width is the image width
+   * @param height is the image height
+   * @param imageFormat is the {@link ImageFormat} that should be returned to dart.
+   * @param maxImages is how many images can be acquired at one time, usually 1.
+   */
+  public ImageStreamReader(int width, int height, int imageFormat, int maxImages) {
+    this.dartImageFormat = imageFormat;
+    this.imageReader =
+        ImageReader.newInstance(width, height, computeStreamImageFormat(imageFormat), maxImages);
+    this.imageStreamReaderUtils = new ImageStreamReaderUtils();
+  }
+
+  /**
+   * Returns the image format to stream based on a requested input format. Usually it's the same
+   * except when dart is requesting NV21. In that case we stream YUV420 and process it into NV21
+   * before sending the frames over.
+   *
+   * @param dartImageFormat is the image format dart is requesting.
+   * @return the image format that should be streamed from the camera.
+   */
+  @VisibleForTesting
+  public static int computeStreamImageFormat(int dartImageFormat) {
+    if (dartImageFormat == ImageFormat.NV21) {
+      return ImageFormat.YUV_420_888;
```

Comment:
Sorry I didn't get it, what's the point supporting NV21, but you change it to YUV? then you convert from YUV to NV21 again? we can do it in dart, and this is not solving the lag of imageStream. Can you enlight us?

Reply:
The reasoning is discussed over in the original PR here: https://github.com/flutter/plugins/pull/6985

Camera2 can't stream NV21, only YUV so if you want it in NV21 you either convert it in dart or on the native side first. To make it easier for others to consume, the `camera_android` package handles the logic for this conversion and handles all of the edge cases such as when the frames contain buffer padding.

The lag you mention is probably just the nature of having to send things over the platform interface, YUV or NV21-converted. There will always be latency here compared to doing something purely on the native side.

Client-side image detection has latency to it even on the native side if you play with the google MLkit example app. The difference is when the latency is on the native side you get latency in the camera preview framerate, so your overlays always match up with the latency and it feels less noticeable. Latency between native and dart means the camera updates faster than the painted UI from Flutter and might feel worse because of that.